[SVG] Faster property access for <rect> and <circle>

mdubet opened the pull request:
#### ce1ed5b53fe6cf9e093f6797798bff4d263fb078
<pre>
[SVG] Faster property access for &lt;rect&gt; and &lt;circle&gt;
<a href="https://bugs.webkit.org/show_bug.cgi?id=273293">https://bugs.webkit.org/show_bug.cgi?id=273293</a>
<a href="https://rdar.apple.com/127085305">rdar://127085305</a>

Reviewed by Simon Fraser, Sam Weinig and Yusuke Suzuki.

This patch avoids looking for the property via the PropertyRegistry hashmap.
It&apos;s a small improvement on MotionMark on Intel.

* Source/WebCore/svg/SVGCircleElement.cpp:
(WebCore::SVGCircleElement::propertyForAttribute const):
* Source/WebCore/svg/SVGCircleElement.h:
* Source/WebCore/svg/SVGElement.h:
* Source/WebCore/svg/SVGRectElement.cpp:
(WebCore::SVGRectElement::propertyForAttribute const):
* Source/WebCore/svg/SVGRectElement.h:
* Source/WebCore/svg/properties/SVGPropertyOwnerRegistry.h:
(WebCore::requires):
(WebCore::SVGPropertyOwnerRegistry::fastAnimatedPropertyLookup):

Canonical link: <a href="https://commits.webkit.org/279304@main">https://commits.webkit.org/279304@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7b17fbb150d5fdad9afacbfc2060365252b9c823

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/53065 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/32402 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/5552 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/56344 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/3788 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/55370 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/39266 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/3515 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/43035 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/2447 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/55163 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/30163 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/45809 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/24163 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/27172 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/3131 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/1947 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/49024 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/3289 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/57939 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/28206 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/3243 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/50430 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/29426 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/46029 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/49731 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/30345 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7800 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/29180 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

weinig commented on Source/WebCore/svg/properties/SVGPropertyOwnerRegistry.h:
```diff
@@ -29,8 +29,15 @@
 #include "SVGAnimatedPropertyPairAccessorImpl.h"
 #include "SVGPropertyAccessorImpl.h"
 #include "SVGPropertyRegistry.h"
+#include <type_traits>
```
Doesn't look like you ended up using this.

weinig commented on Source/WebCore/svg/properties/SVGPropertyOwnerRegistry.h:
```diff
@@ -29,8 +29,15 @@
 #include <wtf/HashMap.h>
 
+template<typename T>
+concept fastPropertyFromAttribute = requires(const T& t)
```
I would call this something like `HasFastPropertyFromAttribute` to make it more consistent with our other meta functions.

Also, very neat!

weinig commented on Source/WebCore/svg/properties/SVGPropertyOwnerRegistry.h:
```diff
@@ -245,10 +255,23 @@ class SVGPropertyOwnerRegistry : public SVGPropertyRegistry {
 
+    static inline SVGAnimatedProperty* fastAnimatedPropertyLookup(OwnerType& owner, const QualifiedName& attributeName)
+    {
+        // if constexpr (std::is_same_v<SVGRectElement, OwnerType> || std::is_same_v<SVGCircleElement, OwnerType>)
```
Looks like you left this in accidentally.

weinig commented on Source/WebCore/svg/properties/SVGPropertyOwnerRegistry.h:
```diff
@@ -245,10 +255,23 @@ class SVGPropertyOwnerRegistry : public SVGPropertyRegistry {
+            return nullptr;
+    }
+
+
```
Not sure you need this much whitespace.

smfr commented on Source/WebCore/svg/SVGCircleElement.h:
```diff
@@ -43,7 +43,9 @@ class SVGCircleElement final : public SVGGeometryElement {
     using PropertyRegistry = SVGPropertyOwnerRegistry<SVGCircleElement, SVGGeometryElement>;
+    friend PropertyRegistry;
 
+    SVGAnimatedProperty* propertyForAttribute(const QualifiedName&);
```
Can the function be `const`?

smfr commented on Source/WebCore/svg/SVGElement.cpp:
```diff
@@ -543,6 +543,15 @@ void SVGElement::attributeChanged(const QualifiedName& name, const AtomString& o
 {
+    /*
+    // Fast known property access, if not found we fallback to the hashmap property registry.
+    if (auto* property = propertyForAttribute(name)) {
```
Why the commented out code?

smfr commented on Source/WebCore/svg/SVGRectElement.cpp:
```diff
@@ -55,6 +56,23 @@ Ref<SVGRectElement> SVGRectElement::create(const QualifiedName& tagName, Documen
     return adoptRef(*new SVGRectElement(tagName, document));
 }
 
+SVGAnimatedProperty* SVGRectElement::propertyForAttribute(const QualifiedName& name)
```
`const`. Can the return value be `const` too?

smfr commented on Source/WebCore/svg/properties/SVGPropertyOwnerRegistry.h:
```diff
@@ -231,6 +237,9 @@ class SVGPropertyOwnerRegistry : public SVGPropertyRegistry {
     void setAnimatedPropertyDirty(const QualifiedName& attributeName, SVGAnimatedProperty& animatedProperty) const override
     {
+        if (auto* property = fastAnimatedPropertyLookup(m_owner, attributeName))
+            return property->setDirty();
```
I find this return pattern confusing.

mdubet replied:
Agreed, this is not pretty. Changing it for 2 separate instructions.

smfr commented on Source/WebCore/svg/properties/SVGPropertyOwnerRegistry.h:
```diff
@@ -245,10 +254,21 @@ class SVGPropertyOwnerRegistry : public SVGPropertyRegistry {
+    {
+        if constexpr (HasFastPropertyFromAttribute<OwnerType>)
+            return owner.propertyForAttribute(attributeName);
+        else
```
No else after return

mdubet replied:
I think it's mandatory because of the `constexpr` in this case

mdubet replied:
Actually no it's not 👍 Changed

Constellation reviewed (approved):
r=me

mdubet committed:
[SVG] Faster property access for <rect> and <circle>
https://bugs.webkit.org/show_bug.cgi?id=273293
rdar://127085305

Reviewed by Simon Fraser, Sam Weinig and Yusuke Suzuki.

This patch avoids looking for the property via the PropertyRegistry hashmap.
It's a small improvement on MotionMark on Intel.

* Source/WebCore/svg/SVGCircleElement.cpp:
(WebCore::SVGCircleElement::propertyForAttribute const):
* Source/WebCore/svg/SVGCircleElement.h:
* Source/WebCore/svg/SVGElement.h:
* Source/WebCore/svg/SVGRectElement.cpp:
(WebCore::SVGRectElement::propertyForAttribute const):
* Source/WebCore/svg/SVGRectElement.h:
* Source/WebCore/svg/properties/SVGPropertyOwnerRegistry.h:
(WebCore::requires):
(WebCore::SVGPropertyOwnerRegistry::fastAnimatedPropertyLookup):

Canonical link: https://commits.webkit.org/279304@main
